eduNEXT/mitx-grading-library
# chore: update actions and test with python 311

## Description

This PR adds support to Python 3.8 and 3.11, drops support for the previous version of Python, and fixes the issues related to this support.

### Changes

- Update the workflows.
- Remove the constraints in the requirements.txt to avoid maintaining unnecessary constraints.
- Change `np.complex` to use `complex` based on a deprecation in numpy 1.20. Ref: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
- Replace `imp` with `importlib` because imp was deprecated in Python 3.4. Ref: https://docs.python.org/3/whatsnew/3.4.html#importlib
- Use approx as the doc suggests. Ref: https://docs.pytest.org/en/4.6.x/reference.html#pytest-approx
- Add `pygments` for the doc, and add `vs-min` as the code theme to maintain the doc `pycon` highlight
- Change the `.asXML()` function because it was deprecated and use `xml` and a custom function. Ref: https://github.com/pyparsing/pyparsing/blob/96d44647450463a7b9c0962b03f3aeb0d1f165d5/CHANGES#L1183
    - First, I tried with `dicttoxml`, which already has the function, but when using the `python_lib.zip` file, Studio doesn't encounter that requirement, and we should add it in the openedx sandbox requirements or add it as a codejail requirement and build the image. I propose using the XML package and creating a function to avoid that path.

## How to test

### Code

- All the checks from Github are passing.

### Integration

**Environment**
- Have a tutor environment with Tutor v18.1.4
- Install codejail. You can follow the installation and usage process from the [tutor-contrib-codejail README](https://github.com/eduNEXT/tutor-contrib-codejail?tab=readme-ov-file#codejail-plugin-for-tutor).
- Perform a dev launch

**Data**
- Import the MIT demo course running:
```
tutor dev do importdemocourse -r https://github.com/mitodl/mitx-grading-library -d course -v master
```
For more info: `tutor dev do importdemocourse -h`
- Create an admin user
```
tutor dev do createuser --staff --superuser admin admin@example.com
```

**Test**
1. Enter in **Studio > MITx Grading Library > Content > Files**
2. Remove the `python_lib.zip` and add this one:
[python_lib.zip](https://github.com/user-attachments/files/17670396/python_lib.zip)
> [!NOTE]  
> If you want to create your python_lib from this version, you can clone the repo, enter in it, and then:
> ```
> find mitxgraders voluptuous -type f ! -path "*/__pycache__/*" | zip python_lib.zip -@
> ```
3. Make sure the units are rendered and worked well.

### Doc

- Create a Python Virtual Environment
- Clone this repo in this branch and enter with `cd`.
- Run `pip install -r requirements.txt`.
- Run `mkdocs serve`.
- Compare the doc with the actual one: https://mitodl.github.io/mitx-grading-library/

## Screencasts

I test the StringGraders, MatrixGraders, and SumGraders in the following video:

[Screencast from 2024-11-07 18-23-57.webm](https://github.com/user-attachments/assets/4003bddc-6c89-4018-b5bb-f520bc2e7481)

## Things checked

- [x] Update basic doc
- [x] Test doc in local
- [x] Test in an environment

## Following work

- Create a PR to remove the Python 2 exceptions and considerations from the code.
- Bump version to 3.0.0 because we have breaking changes and edited the doc.
